### ISSUE TITLE
distinct query problem

### ISSUE THREAD

**Issue by riyase** [realm/realm-java#3537]
#### Goal

> I am having a problem with distinct query. I am getting different results for "distinct()" and "distinctAsync()"
I have five people in my Person table with id, name and age
**id** **name**    **age**
1, Riyas,   20
2, Febin,   21
3, Amogh, 22
4, Jose,     22
5, Mani,     24

when i run the following (not aynchronous) i am getting the expected result
```java
RealmResults<Person> persons = realm.where(Person.class)
                .notEqualTo("age", 20).distinct("age");
```
but when i use asynchonous query it is giving me results with age=20 (that is excluded in query)
```java
RealmResults<Person> persons = realm.where(Person.class)
                .notEqualTo("age", 20).distinctAsync("age");
```

#### Expected Results

> Febin 21
> Amogh 22
> Mani 24

#### Actual Results

> Riyas 20
> Febin 21
> Amogh 22
> Mani 24

#### Code Sample

```java

public class MainActivity extends AppCompatActivity
    implements View.OnClickListener{
    Realm realm;
    TextView tvResult;
    Button bSync;
    Button bAsync;
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        tvResult = (TextView)findViewById(R.id.tvResult);
        bSync = (Button)findViewById(R.id.bSync);
        bSync.setOnClickListener(this);
        bAsync = (Button)findViewById(R.id.bAsync);
        bAsync.setOnClickListener(this);
        setupRealm();
        createDummyPersons();
    }

    private void setupRealm() {
        RealmConfiguration configuration = new RealmConfiguration.Builder(this)
                //.name(REALM_FILE)
                .schemaVersion(0)
                .deleteRealmIfMigrationNeeded()
                .build();
        Realm.setDefaultConfiguration(configuration);
        realm = Realm.getDefaultInstance();
    }

    private void createDummyPersons() {
        realm.executeTransactionAsync(new Realm.Transaction() {
            @Override
            public void execute(Realm realm) {
                Person riyas = getPerson(1,"Riyas",20);
                realm.copyToRealmOrUpdate(riyas);
                Person febin = getPerson(2,"Febin",21);
                realm.copyToRealmOrUpdate(febin);
                Person amogh = getPerson(3,"Amogh",22);
                realm.copyToRealmOrUpdate(amogh);
                Person jose = getPerson(4,"Jose",22);
                realm.copyToRealmOrUpdate(jose);
                Person mani = getPerson(5,"Mani",24);
                realm.copyToRealmOrUpdate(mani);
            }
        });
    }

    private Person getPerson(int id, String name, int age) {
        Person person = new Person();
        person.setId(id);
        person.setName(name);
        person.setAge(age);
        return person;
    }

    private void executeQuery(boolean async) {
        RealmQuery<Person> query = realm.where(Person.class)
                .notEqualTo("age", 20);
        RealmResults<Person> personsResult;
        if (async) {
            personsResult = query.distinctAsync("age");
            personsResult.addChangeListener(new RealmChangeListener<RealmResults<Person>>() {
                @Override
                public void onChange(RealmResults<Person> persons) {
                    persons.removeChangeListeners();
                    String result = createString("Async",persons);
                    tvResult.setText(result);
                }
            });
        } else {
            personsResult = query.distinct("age");
            String result = createString("Sync",personsResult);
            tvResult.setText(result);
        }
    }

    private String createString(String title, RealmResults<Person> persons) {
        StringBuilder builder = new StringBuilder();
        builder.append(title);
        builder.append("\n");
        for (Person person:persons) {
            builder.append(person.getName());
            builder.append(" : ");
            builder.append(person.getAge());
            builder.append("\n");
        }
        return builder.toString();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        realm.close();
    }

    @Override
    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.bSync:
                executeQuery(false);
                break;
            case R.id.bAsync:
                executeQuery(true);
                break;
        }
    }
}



public class Person extends RealmObject {

    @PrimaryKey
    private int id;
    private String name;
    @Index
    private int age;

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public int getAge() {
        return age;
    }

    public void setAge(int age) {
        this.age = age;
    }

    public int getId() {
        return id;
    }

    public void setId(int id) {
        this.id = id;
    }
}


```

#### Version of Realm and tooling
Realm version(s): io.realm:realm-gradle-plugin:2.0.0

Android Studio version: 2.2

**Comment by cmelchior** [comment]
Hi @riyase 
I can reproduce the bug and is looking into it.

**Comment by zaki50** [comment]
fixed in `2.0.1`

**Comment by riyase** [comment]
@cmelchior @zaki50 (y)
